### PR TITLE
#8117: Move global_avg_pool2d to C++

### PR DESCRIPTION
**Pull request by ayerofieiev-tt**
https://github.com/tenstorrent/tt-metal/issues/8117

### CODE REVIEW

File: ttnn/cpp/pybind11/operations/pool.hpp [hunk 1]
```diff
@@ -0,0 +1,67 @@
+// SPDX-FileCopyrightText: © 2023 Tenstorrent Inc.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#pragma once
+
+#include <pybind11/pybind11.h>
+#include <pybind11/stl.h>
+
+#include "../decorators.hpp"
```
Comment (ayerofieiev-tt) [inline ttnn/cpp/pybind11/operations/pool.hpp:10]: copied, will change

File: ttnn/ttnn/operations/pool.py [hunk 2]
```diff
@@ -128,32 +133,8 @@ def _torch_global_avg_pool2d(input_tensor: ttnn.Tensor):
     return torch.nn.functional.global_avg_pool2d(input_tensor, output_size)
 
 
-def _global_avg_pool2d_validate_input_tensors(operation_name, input_tensor, *args, **kwargs):
```
Comment (ayerofieiev-tt) [inline ttnn/ttnn/operations/pool.py:131]: moved to cpp

File: ttnn/ttnn/operations/pool.py [hunk 3]
```diff
@@ -128,32 +133,8 @@ def _torch_global_avg_pool2d(input_tensor: ttnn.Tensor):
...
+    ttnn._ttnn.operations.pool.global_avg_pool2d
 )
-def global_avg_pool2d(input_tensor: ttnn.Tensor, memory_config: ttnn.MemoryConfig = None) -> ttnn.Tensor:
-    r"""
-    Applies a 2D adaptive average pooling over an input signal composed of several input planes.
 
-    Arguments:
-        * :attr: input_tensor: the input tensor
-    """
-    if memory_config is None:
```
Comment (ayerofieiev-tt) [inline ttnn/ttnn/operations/pool.py:155]: behavior now is slightly different, if memory_config is none, we are taking it from the input tensor

Comment (xanderchin) [reply ttnn/ttnn/operations/pool.py:155]: AFAIK, we should be doing DEFAULT_OUTPUT_MEMORY_CONFIG which is in dram interleaved. This is the non-optimal general path that should always work

**Review by arakhmati — APPROVED** [review]

File: ttnn/ttnn/operations/pool.py (hunk 3 above)
Comment (ayerofieiev-tt) [reply ttnn/ttnn/operations/pool.py:155]: We have a thread discussing this. Likely will stay as is, but if needed, will change separately

File: ttnn/cpp/pybind11/operations/__init__.hpp [hunk 4]
```diff
@@ -50,6 +51,9 @@ void py_module(py::module& module) {
 
     auto m_kv_cache = module.def_submodule("kv_cache", "KV cache operations");
     kv_cache::py_module(m_kv_cache);
+
+    auto m_pool = module.def_submodule("pool", "pool operations");
```
[the remark this answers is not in the document]
Comment (ayerofieiev-tt) [reply ttnn/cpp/pybind11/operations/__init__.hpp:55]: Will do 👍

File: ttnn/cpp/ttnn/operations/pool.hpp [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+#include "tt_eager/tt_dnn/op_library/pool/average_pool.hpp"
+
+namespace ttnn {
+namespace operations {
+namespace pool {
+
+namespace detail {
+inline const std::array<ttnn::TensorSchema, 1> input_tensor_schemas() {
+    return {ttnn::TensorSchema{
+        4,  // min rank
```
[the remark this answers is not in the document]
Comment (ayerofieiev-tt) [reply ttnn/cpp/ttnn/operations/pool.hpp:18]: I am not sure, relied on python validation info here. 
I still struggle to understand real limits of the underlying ops from reading them. 
Will gladly take any advice here.

**Pull request merged**